open-webui: init role

osnyx opened the pull request:
@flyingcircusio/release-managers

PL-133951

## Release process

- [x] Created changelog entry using `./changelog.sh`

## PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [ ] ticket state set to *Pull request ready*
- [x] if ticket is more urgent than within the next few days, directly contact a member of the Platform team
<!---->
- [x] set urgency and risk labels
- [ ] ensure the merge bot has determined a merge date
- [ ] ensure all checks are green
- [ ] get a review from a colleague

## Design notes

- [x] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [x] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - integrate authentication system with Flying Circus directory
  - ensure compatibility of role and NixOS upstream module
  - permission model needs to be known by the users
- [x] Security requirements tested? (EVIDENCE)
  - NixOS smoke test for role, ensuring that the NixOS code evaluates and the ExecStart remains compatible
  - manually tested OIDC auth integration with a dev deployment of the directory
  - permission model is documented

osnyx committed:
open-webui: init role

- includes nginx reverse proxy
- by default works against an OpenAI-style web api gateway

PL-133951

osnyx committed:
open-webui: prepare automatic OIDC auth against directory

sysvinit committed:
open-webui: add indirection for loading secrets with systemd LoadCredential

sysvinit committed:
open-webui: remove duplicate configuration

sysvinit committed:
tests: smoke test for open-webui

osnyx committed:
open-webui: expose option for requiring admin approval of new users

osnyx committed:
open-webui: add initial role docs

osnyx committed:
PR changelog

leona-ya commented on nixos/roles/open-webui.nix:
```diff
@@ -0,0 +1,145 @@
+              proxy_http_version 1.1;
+              proxy_set_header Upgrade $http_upgrade;
+              proxy_set_header Connection "upgrade";
+            '';
```
We should be able to simplify this with just having a `/` location and `proxyWebsockets = true`, I think. I think we don't need this special casing

sysvinit replied:
Fixed in [2afee60](https://github.com/flyingcircusio/fc-nixos/pull/1738/commits/2afee60d1d2455e87bf8389a2a70cd41707049c5)

leona-ya commented on nixos/roles/open-webui.nix:
```diff
@@ -0,0 +1,145 @@
+    ];
+
+    services.nginx = {
+      enable = true;
```
This feels very wrong. You're duplicating our default nginx config here, which might cause diversions and conflicts in the future. Instead we should probably use `flyingcircus.services.nginx.enable = true` (or maybe even `flyingcircus.roles.webgateway.enable = true` - I don't know that which one is better).

sysvinit replied:
We adapted this from the code for the statshost role, which uses the same pattern of setting `services.nginx.enable = true` from the depths of the role code. (I believe that webgateway also turns on an haproxy for load balancing which doesn't really seem necessary here.)

I was under the impression that `flyingcircus.services.nginx` is considered deprecated, at least it's documented as such in https://doc.flyingcircus.io/roles/fc-24.11-production/upgrade.html#webgateway-nginx.

leona-ya replied:
That's partly true. We want to deprecate that we carry our own module mostly and use upstream. We will still need a point where we set our defaults - that might be `flyingcircus.roles.nginx` or `flyingcircus.services.nginx`, not sure which one

sysvinit replied:
I've taken a look at `flyingcircus.services.nginx` and we do set quite a few defaults there – the principle of least surprise would speak in favour of using it here as well to ensure consistent behaviour with other nginx instances. I'll discuss this with @osnyx.

sysvinit replied:
Fixed in [2afee60](https://github.com/flyingcircusio/fc-nixos/pull/1738/commits/2afee60d1d2455e87bf8389a2a70cd41707049c5). We've switched to `flyingcircus.services.nginx` to ensure that we have uniform defaults.

leona-ya commented on nixos/roles/open-webui.nix:
```diff
@@ -0,0 +1,145 @@
+    networking.firewall.allowedTCPPorts = [
+      80
+      443
+    ];
```
This can be removed with the change suggested in the comment below

sysvinit replied:
Fixed, see other comments.

leona-ya commented on nixos/roles/open-webui.nix:
```diff
@@ -0,0 +1,145 @@
+    };
+    usersNeedApproval = mkOption {
+      default = false;
+      type = types.boolean;
```
that's not a valid type, it's named `types.bool`

sysvinit replied:
Fixed in ce09f6669e1f3415ae2a7e0d499c1cc698147b66.

leona-ya commented on tests/default.nix:
```diff
@@ -95,6 +95,7 @@ in
   opensearch = callTest ./opensearch.nix { };
   opensearch_dashboards = callTest ./opensearch_dashboards.nix { };
   openvpn = callTest ./openvpn.nix { };
+  open-webui = callTest ./open-webui { };
```
```suggestion
  open-webui = callTest ./open-webui.nix { };
```

sysvinit replied:
Fixed in ce09f6669e1f3415ae2a7e0d499c1cc698147b66.

leona-ya commented on nixos/roles/open-webui.nix:
```diff
@@ -0,0 +1,145 @@
+      default = null;
+      type = types.nullOr types.str;
+      description = "A file containing the authentication secret for the Flying
+      Circus AI Provider.";
```
You probably want to use an actual multiline string here

sysvinit replied:
Fixed in ce09f6669e1f3415ae2a7e0d499c1cc698147b66.

leona-ya commented on nixos/roles/open-webui.nix:
```diff
@@ -0,0 +1,145 @@
+            exec ${lib.getExe scfg.package} serve --host "${scfg.host}" --port ${toString scfg.port}
+          '';
+        in
+        mkOverride 90 "${execStart}/bin/open-webui-start";
```
```suggestion
        mkOverride 90 (lib.getExe execStart);
```

sysvinit replied:
Fixed in ce09f6669e1f3415ae2a7e0d499c1cc698147b66.

sysvinit committed:
open-webui: fix various typos in review feedback

sysvinit committed:
open-webui: switch to flyingcircus nginx module

This ensure that uniform defaults are configured
correctly. Additionally collapse the location configurations into a
single block.

osnyx committed:
open-webui: tune default settings

- disable update check (updates provided by platform)
- enable CORS restrictions

sysvinit committed:
open-webui: refactor config and update defaults

sysvinit committed:
open-webui: fix default upstream api url path

leona-ya commented on doc/src/open-webui.md:
```diff
@@ -0,0 +1,26 @@
+A *Flying Circus AI Services* subscription is required for this role.
+
+Enabling the `open-webui` role automatically sets up Open WebUI available under `https://ai-chat.<resource group name>.fcio.net`. As the first account accessing the application after installation automatically gains administrator privileges, we recommend that a designated admininstrator logs in immediately after setting up the role.
+
```
It would be good to mention the required secret setup process here.

sysvinit replied:
I've added a few sentences in [e42b7b6](https://github.com/flyingcircusio/fc-nixos/pull/1738/commits/e42b7b663aaa33ec33bf20d911463c4fcf6416e6), the details can be adjusted later once we have the other parts all put together.

leona-ya commented on nixos/roles/open-webui.nix:
```diff
@@ -0,0 +1,128 @@
+        ENABLE_OAUTH_SIGNUP = "true";
+        OAUTH_CLIENT_ID = "${config.networking.hostName}_open-webui";
+        OPENID_PROVIDER_URL = "https://auth.flyingcircus.io/realms/fcio/.well-known/openid-configuration";
+        OAUTH_PROVIDER_NAME = "FCIO";
```
We probably want something more accessible for our customers like "Flying Circus"

sysvinit replied:
Fixed in [e42b7b6](https://github.com/flyingcircusio/fc-nixos/pull/1738/commits/e42b7b663aaa33ec33bf20d911463c4fcf6416e6).

leona-ya commented on tests/open-webui.nix:
```diff
@@ -0,0 +1,31 @@
+
+      # open webui is slow to boot, but should eventually respond
+      # with http status 200
+      machine.wait_until_succeeds("curl http://192.168.3.1:8080", timeout=30)
```
The timeout doesn't seem to be long enough, it fails sometimes in Hydra, e.g. https://hydra.flyingcircus.io/build/24410413/nixlog/41/tail

sysvinit replied:
The timeout is due to a change in the configuration since I wrote that test, I've fixed it in [976d3ba](https://github.com/flyingcircusio/fc-nixos/pull/1738/commits/976d3ba125eef6b512a6e85ea78e708d079b16e4).

sysvinit committed:
open-webui: fix test binding to hostname

Ensure that /etc/hosts is configured correctly in the test VM so that
giving a hostname to open-webui as the bind address resolves to the
desired SRV addresses.

sysvinit committed:
open-webui: fixes from review feedback

sysvinit committed:
open-webui: fixes from review feedback
